### ISSUE TITLE
Queries to the bigchain table should ignore invalid blocks

### ISSUE THREAD

**Issue by r-marques** [bigchaindb/bigchaindb#192]
When querying the `bigchain` table the queries should ignore invalid blocks.

related to #134 

TODO:
- [ ] Change read queries to ignore invalid blocks
- [ ] Implement a query to check if a block is valid or not

The queries that we do at this moment are `get_owned_ids`, `get_spent`, `get_transaction`, and `get_tx_by_payload_hash`.

- `get_owned_ids`
Returns the inputs that a given public key owns. We need the query to ignore invalid blocks.
The tricky part is if the block is still undecided. One solution would be to also return the inputs of blocks that are undecided (we can assume that most blocks will be valid). The worst case scenario for this solution is that if the block ends up being invalid, the transaction using the input will also be invalid.

- `get_spent`
Checks if a given txid was already spent. We need the query to ignore invalid blocks (if the input was spent in an invalid block then it was not spent).
Same as above. If the block is undecided we should set it as spent (assume most blocks will be valid)

- `get_transaction`
The query should ignore invalid blocks. Still return if the block is undecided.

- `get_tx_by_payload_hash`
Ignore invalid blocks

For this we also need the ability to query the database for a block and check whether it is undecided or what was the majority decision

**Comment by rhsimplex** [comment]
It seems what we really need is a way to get the block given the transaction.  The current RethinkDB call in `get_transaction` looks like:

```python
    response = r.table('bigchain').concat_map(lambda doc: doc['block']['transactions'])\
        .filter(lambda transaction: transaction['id'] == txid).run(self.conn)
```

The blocks are abstracted away by `concat_map`. Is there an obvious way to preserve the hierarchy in the result (without returning the entire transaction block)? I'll look into it. Would be cool to do this in one DB query rather than two.

Once #221 is merged, we'll have the functionality to check election results. Then it should be easy to query the status of a block: `get_block` could be the analogue of 'get_transaction', and could include a threshold (defaulting to `undecided` or better).

**Comment by r-marques** [comment]
I guess you could filter `bigchain` before the `concat_map`. You can basically run any code on the filter.

something like:
```python
r.table('bigchain').filter(lambda block: b.check_election_status(block) == 'valid').concat_map...
```

**Comment by rhsimplex** [comment]
I tried that, and it doesn't work as-is. Not sure why, perhaps because `b.check_election_status` is an instance method.

**Comment by ttmc** [comment]
PR #324 addresses most of this issue, except (copying from what @rhsimplex wrote on that PR):

"...except for checking blocks on get_tx_by_payload_hash, because no tests exist for this query and I understand we're going to change it soon anyway."

I will close this issue and create a new issue just for that.
